src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+
+  if (!state || state->length_dist.type == CIRCPAD_DIST_NONE) {
+    mi->state_length = CIRCPAD_STATE_LENGTH_INFINITE;
+    return;
+  }
+
+  length = circpad_distribution_sample(state->length_dist);
+  length = MAX(0, length);
+  length += state->start_length;
+  length = MIN(length, state->max_length);
+
+  mi->state_length = clamp_double_to_int64(length);
+}
+
+/**
+ * Sample an expected time-until-next-packet delay from the histogram.
+ *
+ * The bin is chosen with probability proportional to the number
+ * of tokens in each bin, and then a time value is chosen uniformly from
+ * that bin's [start,end) time range.
+ */
+static circpad_delay_t
+circpad_machine_sample_delay(circpad_machineinfo_t *mi)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  const circpad_hist_bin_t *histogram = NULL;
+  int i = 0;
+  uint32_t curr_weight = 0;
+  uint32_t histogram_total = 0;
+  uint32_t bin_choice;
+  circpad_delay_t bin_start, bin_end;
+  circpad_delay_t start_usec;
+
+  tor_assert(state);
+
+  if (state->use_rtt_estimate)
+    start_usec = mi->rtt_estimate_us+state->start_usec;
+  else
+    start_usec = state->start_usec;
+
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    /* We have a mutable histogram */
+    tor_assert(mi->histogram && mi->histogram_len == state->histogram_len);
+
+    histogram = mi->histogram;
+    for (int b = 0; b < state->histogram_len; b++)
+      histogram_total += histogram[b];
+  } else if (state->iat_dist.type == CIRCPAD_DIST_NONE) {
+    /* We have a histogram, but it's immutable */
+    histogram = state->histogram;
+    histogram_total = state->histogram_total;
+  } else {
+    /* Sample from a fixed IAT distribution and return */
+    double val = circpad_distribution_sample(state->iat_dist);
+    /* These comparisons are safe, because the output is in the range
+     * [0, 2**46], and double has a precision of 53 bits. */
+    val = MAX(0, val);
+    val = MIN(val, state->range_sec*TOR_USEC_PER_SEC);
+
+    /* This addition is exact: val is at most 2**16 * USEC_PER_SEC ~= 2**46,
+     * start_usec is at most 2**32, and doubles have a precision of 53 bits. */
+    val += start_usec;
+    /* Clamp the distribution at infinite delay val */
+    return (circpad_delay_t)MIN(tor_llround(val), CIRCPAD_DELAY_INFINITE);
+  }
+
+  bin_choice = crypto_rand_int(histogram_total);
+
+  /* Skip all the initial zero bins */
+  while (!histogram[i]) {
+    i++;
+  }
+  curr_weight = histogram[i];
+
+  // TODO: This is not constant-time. Pretty sure we don't
+  // really need it to be, though.
+  while (curr_weight < bin_choice) {
+    i++;
+    tor_assert(i < state->histogram_len);
+    curr_weight += histogram[i];
+  }
+
+  tor_assert(i < state->histogram_len);
+  tor_assert(histogram[i] > 0);
+
+  // Store this index to remove the token upon callback.
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    mi->chosen_bin = i;
+  }
+
+  if (i == state->histogram_len-1) {
```

Comment:
I wonder if we can make it more clear that `histogram_len-1` is the infinity bin, since sometimes it's not clear that this is what we are doing.

There are various places in the code that compute `histogram_len-1` and act on it. Perhaps we can introduce a function called `get_infinity_bin_index()` or something?

Reply:
5b39c9d3d8ffe75cb5a5fbcc5f9f23c8a2ef1362 and ed6284a41b6dca65bd2b19c2f2a066802beb10d2. Macro is better because it will work with circpad_state_t and circpad_machineinfo_t (the infinity bin checks apply to both of those, and the fields have the same name).